taraxa/state/dpos/precompiled/dpos_contract.go

```diff
@@ -0,0 +1,1022 @@
+	if state == nil {
+		old_state := self.state_get_and_decrement(args.Validator[:], BlockToBytes(validator.LastUpdated))
+		state = new(State)
+		state.RwardsPer1Stake = bigutil.Add(old_state.RwardsPer1Stake, bigutil.Div(validator.RewardsPool, validator.TotalStake))
+		// TODO: question: how can we erase validator's RewardsPool during delegation of single delegator ???
+		validator.RewardsPool = bigutil.Big0
+		validator.LastUpdated = block
+		state.Count++
+	}
+
+	delegation := self.delegations.GetDelegation(ctx.CallerAccount.Address(), &args.Validator)
+	if delegation == nil {
+		ctx.Account.SubBalance(ctx.Value)
+		self.delegations.CreateDelegation(ctx.CallerAccount.Address(), &args.Validator, block, ctx.Value)
+		validator.TotalStake = bigutil.Add(validator.TotalStake, ctx.Value)
+	} else {
+		// We need to claim rewards first
+		old_state := self.state_get_and_decrement(args.Validator[:], BlockToBytes(delegation.LastUpdated))
+		reward := bigutil.Sub(state.RwardsPer1Stake, old_state.RwardsPer1Stake)
+		ctx.CallerAccount.AddBalance(bigutil.Mul(reward, delegation.Stake))
+
+		ctx.Account.SubBalance(ctx.Value)
+
+		delegation.Stake = bigutil.Add(delegation.Stake, ctx.Value)
+		delegation.LastUpdated = block
+		self.delegations.ModifyDelegation(ctx.CallerAccount.Address(), &args.Validator, delegation)
+
+		validator.TotalStake = bigutil.Add(validator.TotalStake, ctx.Value)
+	}
+
+	self.amount_delegated = bigutil.Add(self.amount_delegated, ctx.Value)
+	eligible_now := validator.TotalStake.Cmp(self.cfg.EligibilityBalanceThreshold) >= 0
+	if !was_eligible && eligible_now {
+		self.eligible_count++
+	}
+	new_vote_count := vote_count(validator.TotalStake, self.cfg.EligibilityBalanceThreshold, self.cfg.VoteEligibilityBalanceStep)
+	if prev_vote_count != new_vote_count {
+		self.eligible_vote_count -= prev_vote_count
+		self.eligible_vote_count = Add64p(self.eligible_vote_count, new_vote_count)
+	}
+
+	state.Count++
+	self.state_put(&state_k, state)
+	self.validators.ModifyValidator(&args.Validator, validator)
+	return nil
+}
+
+func (self *Contract) undelegate(ctx vm.CallFrame, block types.BlockNum, args UndelegateArgs) error {
+	if self.undelegations.UndelegationExists(ctx.CallerAccount.Address(), &args.Validator) {
+		return ErrExistentUndelegation
+	}
+
+	validator := self.validators.GetValidator(&args.Validator)
+	if validator == nil {
+		return ErrNonExistentValidator
+	}
+	was_eligible := validator.TotalStake.Cmp(self.cfg.EligibilityBalanceThreshold) >= 0
+	prev_vote_count := vote_count(validator.TotalStake, self.cfg.EligibilityBalanceThreshold, self.cfg.VoteEligibilityBalanceStep)
+
+	delegation := self.delegations.GetDelegation(ctx.CallerAccount.Address(), &args.Validator)
+	if delegation == nil {
+		return ErrNonExistentDelegation
+	}
+
+	if delegation.Stake.Cmp(args.Amount) == -1 {
+		return ErrInsufficientDelegation
+	}
+
+	state, state_k := self.state_get(args.Validator[:], BlockToBytes(block))
```

Comment:
Missing check where user must have min. 1000 tokens staked. Lets say he has 1500 staked and want to undelegate 1000. We should not let him do that. He can either undelegate <1, 500> or everything -> 1500

Reply:
as I said before all those values need to be specified in some task, where everything is documented. Current approach copies old DPOS

Reply:
https://docs.google.com/document/d/16xodJLO2LePq-EZVZhPw07pZwVwpD5E-sch3pAevdwg/edit